pkg/aws/client/client.go

```diff
@@ -418,6 +423,90 @@ func (c *Client) DeleteSecurityGroup(ctx context.Context, id string) error {
 	return ignoreNotFound(err)
 }
 
+// GetDNSHostedZones returns a map of all DNS hosted zone names mapped to their IDs.
+func (c *Client) GetDNSHostedZones(ctx context.Context) (map[string]string, error) {
+	zones := make(map[string]string)
+	if err := c.Route53.ListHostedZonesPagesWithContext(ctx, &route53.ListHostedZonesInput{}, func(out *route53.ListHostedZonesOutput, lastPage bool) bool {
+		for _, zone := range out.HostedZones {
+			zones[normalizeZoneName(aws.StringValue(zone.Name))] = normalizeZoneId(aws.StringValue(zone.Id))
+		}
+		return !lastPage
+	}); err != nil {
+		return nil, err
+	}
+	return zones, nil
+}
+
+func normalizeZoneName(zoneName string) string {
+	if strings.HasPrefix(zoneName, "\\052.") {
+		zoneName = "*" + zoneName[4:]
+	}
+	if strings.HasSuffix(zoneName, ".") {
+		return zoneName[:len(zoneName)-1]
+	}
+	return zoneName
+}
+
+func normalizeZoneId(zoneId string) string {
+	parts := strings.Split(zoneId, "/")
+	return parts[len(parts)-1]
+}
+
+// CreateOrUpdateDNSRecord creates or updates the DNS record in the DNS hosted zone with the given zone ID,
+// with the given name, type, values, and TTL.
+func (c *Client) CreateOrUpdateDNSRecord(ctx context.Context, zoneId, name, recordType string, values []string, ttl int64) error {
+	out, err := c.Route53.ChangeResourceRecordSetsWithContext(ctx, newChangeResourceRecordSetsInput(zoneId, route53.ChangeActionUpsert, name, recordType, values, ttl))
+	if err != nil {
+		return err
+	}
+	return c.Route53.WaitUntilResourceRecordSetsChangedWithContext(ctx, &route53.GetChangeInput{Id: out.ChangeInfo.Id})
+}
+
+// DeleteDNSRecord deletes the DNS record in the DNS hosted zone with the given zone ID,
+// with the given name, type, values, and TTL.
+func (c *Client) DeleteDNSRecord(ctx context.Context, zoneId, name, recordType string, values []string, ttl int64) error {
+	out, err := c.Route53.ChangeResourceRecordSetsWithContext(ctx, newChangeResourceRecordSetsInput(zoneId, route53.ChangeActionDelete, name, recordType, values, ttl))
+	if err != nil {
+		return ignoreNotFoundRoute53(err)
+	}
+	return c.Route53.WaitUntilResourceRecordSetsChangedWithContext(ctx, &route53.GetChangeInput{Id: out.ChangeInfo.Id})
+}
+
+func newChangeResourceRecordSetsInput(zoneId, action, name, recordType string, values []string, ttl int64) *route53.ChangeResourceRecordSetsInput {
+	var resourceRecords []*route53.ResourceRecord
+	for _, value := range values {
+		if recordType == route53.RRTypeTxt {
+			value = encloseInQuotes(value)
+		}
+		resourceRecords = append(resourceRecords, &route53.ResourceRecord{
+			Value: aws.String(value),
+		})
```

Comment:
Load balancers in AWS are using domain names instead of IP addresses. Therefore `CNAME` records need to be supported (and converted into `ALIAS` target records typically). But as far as I can see, the implementation already supports it, as the record type is not validated. `CNAME` records can only have a single value, but an explicit validation is probably not needed as the AWS route53 service will report an error.

Reply:
Thanks a lot! So if I understand correctly, in this case:
* The `RecordType` of the DNSRecord in this case should be `CNAME`. We'll add explicit validation for only a single value in this case (as well as for other things, e.g. the supported values for `recordType`) to https://github.com/gardener/gardener/pull/4231 before merging it.
* Only for the `aws-route53` implementation, this can be converted to an "alias target" using the code you referred to above as an example. 

Question: What is the advantage of using "alias target" instead of a regular `CNAME`? Could we simply live with `CNAME` everywhere?

Reply:
OK, I found https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-choosing-alias-non-alias.html and this somewhat answers my last question. Perhaps the important one is "Pricing for DNS queries" (e.g. alias records are "free" and CNAME are not). Are there any other considerations?

Can the hardcoded list https://github.com/gardener/external-dns-management/blob/2b72ef8cb031137d2945b8253a27dcb1042e733d/pkg/controller/provider/aws/aliastarget.go#L29 be avoided? How likely is it change and to have to be actively maintained?

Reply:
You guessed right. The usage of alias was requested from the SAP Hana team to reduce costs for DNS queries.

We have found no other solution than to hardcode the list. There seems to be no API to fetch this list. We even considered scraping HTML from the page [Elastic Load Balancing endpoints and quotas](https://docs.aws.amazon.com/general/latest/gr/elb.html), but this is of course not robust. The assumption is that the zone IDs never change. If there are new AWS regions we may not be able to generate `ALIAS` as long as they are not appended to the list.

Reply:
OK, thanks, will do as suggested then.

Reply:
Done.